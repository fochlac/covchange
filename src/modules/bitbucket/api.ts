import * as request from 'request-promise-native'

import error from '../../utils/error'

const bitbucketurl = `${global.bitbucket}/rest/api/1.0/`
const Authorization = global.bitbucketKey
const { internalError } = error('Bitbucket API call failed: ')

export const Bitbucket = {
	get: (url: string): Promise<any> => apiCall({ url }),
	put: (url: string, body: any): Promise<any> => apiCall({ method: 'put', url, body }),
	post: (url: string, body: any): Promise<any> => apiCall({ method: 'post', url, body }),
}

interface ApiConfig {
	uri: string
	method: string
	json: boolean
	headers: { [key: string]: string }
	body?: any
}
interface ApiCallParameter {
	url: string
	method?: string
	body?: any
}
function apiCall({ url, method, body }: ApiCallParameter): Promise<any> {
	const config: ApiConfig = {
		uri: bitbucketurl + url,
		method,
		json: true,
		headers: {
			Authorization,
		},
	}

	if (body) {
		config.body = body
	}

	return request(config).catch(internalError(2, 'Error posting comment.'))
}
