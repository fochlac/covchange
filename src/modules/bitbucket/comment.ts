import * as request from 'request-promise-native'

import { commentDb } from '../db/comments'
import { diffReports } from '../diff/diff'
import error from '../../utils/error'

const { internalError } = error('bitbucket comment:')

const Authorization = 'Bearer OTQyODM5NTkyMzQ2OqjtV+lhUZPQdkGwO7kqarTLMxrB'
const baseUrl = ({ repository: { slug, name }, pullRequestId }: Core.PullRequest) =>
	`${global.bitbucket}/rest/api/1.0/${slug}/repos/${name}/pull-requests/${pullRequestId}/comments/`

const symbolsMap = {
	critical: ':u7981:',
	low: ':u6709:',
	medium: ':symbols:',
	high: ':u6307:',
}

const getSymbolFromCoverage = (metrics: Core.Metrics) => {
	const { statementCov, statements } = metrics
	const stringcode = statementCov > 85 ? 'high' : statementCov > 70 || statements <= 5 ? 'medium' : statementCov > 50 ? 'low' : 'critical'
	return symbolsMap[stringcode]
}

export function createComment(base: Core.Branch, pr: Core.PullRequest) {
	const diffLatest = diffReports(base.reports[0], pr.reports[0])

	return sendComment(pr, createCommentObject(diffLatest))
}

async function sendComment(pr, comment) {
	const existingComment = await commentDb.get(pr.repository, pr.pullRequestId)
	const uri = baseUrl(pr) + (existingComment ? existingComment.commentId : '')
	const method = existingComment ? 'PUT' : 'POST'

	return request({
		uri,
		method,
		json: true,
		headers: {
			Authorization,
		},
		body: comment,
	}).catch(internalError(2, 'Error posting comment.'))
}

function createCommentObject(diff: Core.DiffReport) {
	const totalDiff = diff.total.diff.statementCov
	const changedFilesHeader = (Object.keys(diff.changed).length && ['| File | Change | Coverage |', '|---|---|---|']) || []
	const deletedFilesHeader = (diff.deleted.length && ['##### Deleted files', '```diff']) || []
	const deletedFilesFooter = (diff.deleted.length && ['```']) || []

	const lines = [
		'### Coverage Statistics',
		`#### ${totalDiff <= 0 ? ':red_circle:' : ':large_blue_circle:'} This pull request will ` +
			`${totalDiff > 0 ? 'increase' : 'decrease'} total coverage by ${Math.abs(totalDiff)}% to ${diff.total.changed.statementCov}%.`,
		'',
		...changedFilesHeader,
		...Object.keys(diff.changed)
			.sort((a, b) => diff.changed[a].diff.statementCov - diff.changed[b].diff.statementCov)
			.slice(0, 10)
			.map(name => changedFileInfo(name, diff.changed[name])),
		...Object.keys(diff.new)
			.sort((a, b) => diff.new[a].statementCov - diff.new[b].statementCov)
			.slice(0, 10)
			.map(name => newFileInfo(name, diff.new[name])),
		...deletedFilesHeader,
		...diff.deleted.map(deletedFileInfo),
		...deletedFilesFooter,
	]

	return { text: lines.join('\n') }
}

function newFileInfo(name: string, metrics: Core.Metrics) {
	return `| ${name} | :sparkles: *new* | ${getSymbolFromCoverage(metrics)} ${metrics.statementCov}% |`
}

function deletedFileInfo(name: string) {
	return `- ${name}`
}

function changedFileInfo(name: string, diff: Core.Diff) {
	const covSymbol = diff.diff.statementCov < 0 ? ':red_circle:' : ':large_blue_circle:'
	return `| ${name} | ${covSymbol} ${diff.diff.statementCov}% | ${getSymbolFromCoverage(diff.changed)} ${diff.changed.statementCov}% |`
}
