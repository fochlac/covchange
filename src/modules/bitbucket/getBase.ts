import * as request from 'request-promise-native'

import error from '../../utils/error'

const { internalError } = error('bitbucket comment:')

const Authorization = global.bitbucketKey
const baseUrl = ({ repository: { slug, name }, pullRequestId }: Core.PullRequestRest) =>
	`${global.bitbucket}/rest/api/1.0/${slug}/repos/${name}/pull-requests/${pullRequestId}/`


export async function getBaseBranchFromPullrequest(pullRequest:Core.PullRequestRest):Promise<Core.BaseBranch> {
	const uri = baseUrl(pullRequest)
	try {
		const { toRef } = await request({
			uri,
			method: 'GET',
			json: true,
			headers: {
				Authorization,
			}
		})

		const slug = toRef.repository.project.type === 'NORMAL' ? 'projects' : 'users'

		return {
			name: toRef.displayId,
			repository: {
				name: toRef.repository.slug,
				slug: `${slug}/${toRef.repository.project.key}`
			}
		}

	}
	catch(err) {
		internalError(2, 'Error fetching comment.')(err)
	}
}
