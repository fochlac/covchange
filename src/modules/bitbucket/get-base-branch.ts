import { Bitbucket } from './bitbucket'

const baseUrl = ({ repository: { type, repo, project }, name }: Core.PullRequestRest) =>
	`${type}/${project}/repos/${repo}/pull-requests/${name}/`

export async function getBaseBranchFromPullrequest(pullRequest: Core.PullRequestRest): Promise<Core.BaseBranch> {
	const { toRef } = await Bitbucket.get(baseUrl(pullRequest))

	const type = toRef.repository.project.type === 'NORMAL' ? 'projects' : 'users'

	return {
		name: toRef.displayId,
		repository: {
			repo: toRef.repository.slug,
			type,
			project: toRef.repository.project.key,
		},
	}
}
