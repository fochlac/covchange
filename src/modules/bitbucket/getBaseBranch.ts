import * as request from 'request-promise-native'

import { Bitbucket } from './bitbucket'
import error from '../../utils/error'

const { internalError } = error('bitbucket comment:')

const baseUrl = ({ repository: { type, repo, project }, name }: Core.PullRequestRest) =>
	`${type}/${repo}/repos/${project}/pull-requests/${name}/`

export async function getBaseBranchFromPullrequest(pullRequest: Core.PullRequestRest): Promise<Core.BaseBranch> {
	try {
		const { toRef } = await Bitbucket.get(baseUrl(pullRequest))

		const type = toRef.repository.project.type === 'NORMAL' ? 'projects' : 'users'

		return {
			name: toRef.displayId,
			repository: {
				repo: toRef.repository.slug,
				type,
				project: toRef.repository.project.key,
			},
		}
	} catch (err) {
		internalError(2, 'Error fetching comment.')(err)
	}
}
