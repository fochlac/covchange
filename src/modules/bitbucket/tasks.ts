import { Bitbucket } from './bitbucket'

function createApiSlug(taskId?: Core.Id) {
	return `tasks/${taskId || ''}`
}

function updateTask(comment: Core.Comment, open?: boolean): Promise<Core.Id> {
	return Bitbucket.put(createApiSlug(comment.taskId), {
		id: comment.taskId,
		state: open ? 'OPEN' : 'RESOLVED',
	}).then(({ id }: { id: number }) => id.toString())
}

function createTask(comment: Core.Comment): Promise<Core.Id> {
	return Bitbucket.post(createApiSlug(), {
		text: 'Please add more unit tests.',
		anchor: {
			id: comment.commentId,
			type: 'COMMENT',
		},
	}).then(({ id }: { id: number }) => id.toString())
}

export function openTask(comment: Core.Comment): Promise<Core.Id> {
	return comment.taskId ? updateTask(comment, true) : createTask(comment)
}

export function resolveTask(comment: Core.Comment) {
	return comment.taskId ? updateTask(comment, false) : Promise.resolve(comment.taskId)
}
