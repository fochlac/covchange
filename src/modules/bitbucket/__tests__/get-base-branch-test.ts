import { Bitbucket } from '../bitbucket'
import { getBaseBranchFromPullrequest } from '../get-base-branch'

jest.mock('../../../utils/logger.ts', () => () => null)
jest.mock('../bitbucket', () => ({
	Bitbucket: {
		get: jest.fn(),
	},
}))

describe('get-base-branch', () => {
	const toRef = type => ({
		displayId: 'name',
		repository: { slug: 'repo', project: { key: 'project', type } },
	})

	const repository = {
		repo: 'name',
		project: 'slug',
		type: 'users',
	}
	const pr = {
		repository,
		name: '12345',
		report: null,
	}

	afterEach(() => {
		// @ts-ignore
		Bitbucket.get.mockClear()
	})

	it('should create a normal project type branch', () => {
		// @ts-ignore
		Bitbucket.get.mockImplementation(() => Promise.resolve({ toRef: toRef('NORMAL') }))
		return getBaseBranchFromPullrequest(pr).then(baseBranch => {
			expect(Bitbucket.get).toBeCalledWith('users/slug/repos/name/pull-requests/12345/')
			expect(baseBranch).toMatchSnapshot()
		})
	})

	it('should create a normal project type branch', () => {
		// @ts-ignore
		Bitbucket.get.mockImplementation(() => Promise.resolve({ toRef: toRef('USER') }))
		return getBaseBranchFromPullrequest(pr).then(baseBranch => {
			expect(Bitbucket.get).toBeCalledWith('users/slug/repos/name/pull-requests/12345/')
			expect(baseBranch).toMatchSnapshot()
		})
	})
})
