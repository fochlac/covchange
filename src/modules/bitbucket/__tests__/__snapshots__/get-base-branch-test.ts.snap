// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-base-branch should create a normal project type branch 1`] = `
Object {
  "name": "name",
  "repository": Object {
    "project": "project",
    "repo": "repo",
    "type": "projects",
  },
}
`;

exports[`get-base-branch should create a normal project type branch 2`] = `
Object {
  "name": "name",
  "repository": Object {
    "project": "project",
    "repo": "repo",
    "type": "users",
  },
}
`;
